Services/DevicePropertiesRequest.cs

```diff
@@ -0,0 +1,91 @@
+    {
+        Task RegisterDevicePropertyUpdatesAsync(
+            string deviceId,
+            ISmartDictionary deviceProperties);
+
+        Task OnPropertyUpdateRequestedCallback(
+            TwinCollection desiredProperties,
+            object userContext);
+    }
+
+    public class DevicePropertiesRequest : IDevicePropertiesRequest
+    {
+        private readonly Azure.Devices.Client.DeviceClient client;
+        private readonly ILogger log;
+        private string deviceId;
+        private ISmartDictionary deviceProperties;
+
+        public DevicePropertiesRequest(Azure.Devices.Client.DeviceClient client, ILogger logger)
+        {
+            this.client = client;
+            this.log = logger;
+            this.deviceId = string.Empty;
+        }
+
+        public async Task RegisterDevicePropertyUpdatesAsync(string deviceId, ISmartDictionary deviceProperties)
+        {
+            if (this.deviceId != string.Empty)
+            {
+                this.log.Error("Application error, each device must have a separate instance", () => { });
+                throw new Exception("Application error, each device must have a separate instance of " + this.GetType().FullName);
+            }
+
+            this.deviceId = deviceId;
+            this.deviceProperties = deviceProperties;
+
+            this.log.Debug("Setting up callback for desired properties updates.", () => new { this.deviceId });
+
+            // Set callback that IoT Hub calls whenever the client receives a desired properties state update.
+            await this.client.SetDesiredPropertyUpdateCallbackAsync(OnPropertyUpdateRequestedCallback, null);
+
+            this.log.Debug("Callback for desired properties updates setup successfully", () => new { this.deviceId });
+        }
+
+        /// <summary>
+        /// When a desired property change is requested, update the internal the device state properties
+        /// which will be reported to the hub. If there is a new desired property that does not exist in
+        /// the reported properties, it will be added.
+        /// </summary>
+        public Task OnPropertyUpdateRequestedCallback(TwinCollection desiredProperties, object userContext)
+        {
+            this.log.Info("Desired property update requested", () => new { this.deviceId, desiredProperties });
+
+            // This is where custom code for handling specific desired property changes could be added.
+            // For the purposes of the simulation service, we have chosen to write the desired properties
+            // directly to the reported properties. 
+            try
+            {
+                foreach (KeyValuePair<string, object> item in desiredProperties)
+                {
+                    // Only update if key doesn't exist or value has changed 
+                    if (!this.deviceProperties.Has(item.Key) ||
+                        (item.Value.ToString() != this.deviceProperties.Get(item.Key).ToString()))
```

Comment:
wondering if `ToString` is the right way of comparing values. For instance, what if the value is a complex object? Does `ToString` serialize the actual value?

Reply:
I used it because it is the comparison type used in the previous implementation that was proven to work:
https://github.com/Azure/device-simulation-dotnet/blob/0d37f326717db065dab767105763d29a0ece24bd/SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs#L160

However, ToString is overall not very efficient for comparison because every property's value is accessed. I thought of possibly implementing a comparer but because the objects are generic, couldn't think of a better way to go about it. 

Any alternate suggestions?

Reply:
I think that "every value is accessed" is a requirement in order to compare the value, right?

In this before-vs-after example, to detect that "five" was added under the "someproperty" property,
the code needs to see all the props:

```
"desired-properties": {
	"someproperty": {
	    "one": {
	       "two": 2,
	       "three": 3
	    },
	    "four": [ 0,1,2,3 ]
	}
}
```

```
"desired-properties": {
	"someproperty": {
	    "one": {
	       "two": 2,
	       "three": 3,
	       "five": 5
	    },
	    "four": [ 0,1,2,3 ]
	}
}
```

Reply:
Since this meets the parity requirements for RMv2 Preview, I've filed a bug here that we can address:
https://github.com/Azure/device-simulation-dotnet/issues/194